src/hooks.rs

```diff
@@ -0,0 +1,45 @@
+//! This module provides a way to register decoding hooks for image formats not directly supported
+//! by this crate.
+
+use std::{
+    collections::HashMap,
+    io::{BufReader, Read, Seek},
+    sync::RwLock,
+};
+
+use crate::{ImageDecoder, ImageFormat, ImageResult};
+
+pub(crate) trait ReadSeek: Read + Seek {}
+impl<T: Read + Seek> ReadSeek for T {}
+
+pub(crate) static DECODING_HOOKS: RwLock<Option<HashMap<ImageFormat, DecodingHook>>> =
+    RwLock::new(None);
+
+/// A wrapper around a type-erased trait object that implements `Read` and `Seek`.
+pub struct BoxReadSeek<'a>(pub(crate) Box<dyn ReadSeek + 'a>);
+impl Read for BoxReadSeek<'_> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+impl Seek for BoxReadSeek<'_> {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        self.0.seek(pos)
+    }
+}
+
+/// A function to produce an `ImageDecoder` for a given image format.
+pub type DecodingHook = Box<
+    dyn for<'a> Fn(BufReader<BoxReadSeek<'a>>) -> ImageResult<Box<dyn ImageDecoder + 'a>>
+        + Send
+        + Sync,
+>;
+
+/// Register a new decoding hook or replace an existing one.
+pub fn register_decoding_hook(format: ImageFormat, hook: DecodingHook) {
+    let mut hooks = DECODING_HOOKS.write().unwrap();
+    if hooks.is_none() {
+        *hooks = Some(HashMap::new());
+    }
+    hooks.as_mut().unwrap().insert(format, hook);
+}
```

Comment:
I like the internal mechanism, but am not completely sure about the interface. It seems primarily designed around the implementation instead of user expectations. For instance, if we consider the case of third-party support for a decoder that isn't enabled by feature switches then we'd expect both decoder and encoder in a single line.

It's possible to register a hook that's ignored, and that ignore can happen by a feature flag enabled via a sibling create. This poses the question, should there be a mechanism for discovering the utilized features and should the internal translation pre-register itself in the map to provide a more uniform view?

Reply:
Registering both encoding and decoding in one go can always be added later as a dedicated function. A 3rd party codec crate can always provide their method to register themselves, like `my_codec_crate::register()`.

But it could be common to want *different* implementations for decoding and encoding. For decoding you may want a decoder that is safe and/or compatible, but for encoding you may prefer either a fast one or heavily-compressing one (safety is usually less of an issue when encoding).

Reply:
Yeah, I was kind of imagining an external crate would expose its own `register()` function that registered all the encoders/decoders it had. That could be only a single decoder, or several encoders and decoders all defined in the same crate 

I didn't add encoding support to this PR because at the moment `Box<dyn ImageEncoder>` isn't actually usable for encoding images (you can't call `write_image` because that requires `Self: Sized`).